game/songs.cc

```diff
@@ -82,181 +87,160 @@ void Songs::reload_internal() {
+		}
+		if(!std::isnan(song->preview_start)) {
+			songObject["previewStart"] = song->preview_start;
+		}
+		if(!song->music[TrackName::LEAD_VOCAL].string().empty()) {
+			songObject["vocals"] = song->music[TrackName::LEAD_VOCAL].string();
+		}
+		if(!song->music[TrackName::PREVIEW].string().empty()) {
+			songObject["preview"] = song->music[TrackName::PREVIEW].string();
+		}
+		if(!song->music[TrackName::GUITAR].string().empty()) {
+			songObject["guitar"] = song->music[TrackName::GUITAR].string();
+		}
+		if(!song->music[TrackName::BASS].string().empty()) {
+			songObject["bass"] = song->music[TrackName::BASS].string();
+		}
+		if(!song->music[TrackName::DRUMS].string().empty()) {
+			songObject["drums"] = song->music[TrackName::DRUMS].string();
+		}
+		if(!song->music[TrackName::KEYBOARD].string().empty()) {
+			songObject["keyboard"] = song->music[TrackName::KEYBOARD].string();
+		}
+		if(!song->music[TrackName::GUITAR_COOP].string().empty()) {
+			songObject["guitarCoop"] = song->music[TrackName::GUITAR_COOP].string();
+		}
+		if(!song->music[TrackName::GUITAR_RHYTHM].string().empty()) {
+			songObject["guitarRhythm"] = song->music[TrackName::GUITAR_RHYTHM].string();
+		}
+
+		double duration = song->getDurationSeconds();
+		if(!std::isnan(duration)) {
+			songObject["duration"] = duration;
+		}
+		if (!song->m_bpms.empty()) {
+			songObject["bpm"] = 15 / song->m_bpms.front().step;
+		}
+
 		// Cache songtype also.
 		if(song->hasVocals()) {
-			uint32_t vocals = song->vocalTracks.size();
-        	songObject["VocalTracks"] = web::json::value::number(vocals);
-    	}
-		if(song->hasKeyboard()) {
-        	songObject["KeyboardTracks"] = web::json::value::number(1);
-    	}
-		if(song->hasDrums()) {
-        	songObject["DrumTracks"] = web::json::value::number(1);
-    	}
-		if(song->hasDance()) {
-			uint32_t dance = song->danceTracks.size();
-        	songObject["DanceTracks"] = web::json::value::number(dance);
-    	}
-		if(song->hasGuitars()) {
-			uint32_t guitars = song->instrumentTracks.size() - song->hasDrums() - song->hasKeyboard();
-        	songObject["GuitarTracks"] = web::json::value::number(guitars);
-    	}
-	    if(songObject != web::json::value::object()) {
-        	jsonRoot[i] = songObject;
-        	i++;
-    	}
+			songObject["vocalTracks"] = song->vocalTracks.size();
+		}
+		songObject["keyboardTracks"] = song->hasKeyboard();
+		songObject["drumTracks"] = song->hasDrums();
+		songObject["danceTracks"] = song->hasDance();
+		songObject["guitarTracks"] = song->hasGuitars();
+		if(songObject != nlohmann::json::object()) {
+			jsonRoot.push_back(songObject);
+		}
 	}
 
-	fs::path cacheDir = getCacheDir() / "Songs-Metadata.json";
+	fs::path cacheDir = getCacheDir() / SONGS_CACHE_JSON_FILE;
 
 	try {
-		std::stringstream stream;
-		jsonRoot.serialize(stream);
 		std::ofstream outFile(cacheDir.string());
-    	outFile << stream.rdbuf();
-    	outFile.close();
+		outFile << jsonRoot.dump(4);
```

Comment:
Define the 'magic number' by using a const variable.
`const int spacesCount = 4` or a better name.